tokio/src/sync/watch.rs

```diff
@@ -610,6 +610,93 @@ impl<T> Receiver<T> {
+    /// closure, then it may skip some updates. Whenever the closure is called,
+    /// it will be called with the most recent value.
+    ///
+    /// When this function returns, the value that was passed to the closure
+    /// when it returned `true` will be considered seen.
+    ///
+    /// If the channel is closed, then `wait_for` will return a `RecvError`.
+    /// Once this happens, no more messages can ever be sent on the channel.
+    /// When an error is returned, it is guaranteed that the closure has been
+    /// called on the last value, and that it returned `false` for that value.
+    /// (If the closure returned `true`, then the last value would have been
+    /// returned instead of the error.)
+    ///
+    /// Like the `borrow` method, the returned borrow holds a read lock on the
+    /// inner value. This means that long-lived borrows could cause the producer
+    /// half to block. It is recommended to keep the borrow as short-lived as
+    /// possible. See the documentation of `borrow` for more information on
+    /// this.
+    ///
+    /// [`Receiver::changed()`]: crate::sync::watch::Receiver::changed
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::watch;
+    ///
+    /// #[tokio::main]
+    ///
+    /// async fn main() {
+    ///     let (tx, _rx) = watch::channel("hello");
+    ///
+    ///     tx.send("goodbye").unwrap();
+    ///     
+    ///     // here we subscribe to a second receiver
+    ///     // now in case of using `changed` we would have
+    ///     // to first check the current value and then wait
+    ///     // for changes or else `changed` would hang.
+    ///     let mut rx2 = tx.subscribe();
+    ///     
+    ///     // in place of changed we have use `wait_for`
+    ///     // which would automatically check the current value
+    ///     // and wait for changes until the closure returns true.
+    ///     assert!(rx2.wait_for(|val| *val == "goodbye").await.is_ok());
+    ///     assert_eq!(*rx2.borrow(), "goodbye");
+    /// }
+    /// ```
+    pub async fn wait_for(
+        &mut self,
+        mut f: impl FnMut(&T) -> bool,
+    ) -> Result<Ref<'_, T>, error::RecvError> {
+        loop {
+            if f(&self.borrow_and_update()) {
+                return Ok(self.borrow());
+            }
+
+            match self.changed().await {
+                Ok(_) => {}
+                Err(e) => {
+                    // check if the version has changed since the last time we
+                    // called the closure.
+                    if self.shared.state.load().version() == self.version {
+                        return Err(e);
+                    }
+                    // the version has changed, so we need to call the closure
+                    if f(&self.borrow_and_update()) {
+                        return Ok(self.borrow());
+                    }
+                    return Err(e);
```

Comment:
It seems like you can handle this using the `has_changed` field in the `Ref` returned by `borrow_and_update`.
```Rust
let val = self.borrow_and_update();
if val.has_changed && f(&val) {
    return Ok(val);
}
return Err(e);
```

Reply:
updated in 
09652f8